streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java

```diff
@@ -222,6 +223,29 @@ synchronized void unlock(final TaskId taskId) throws IOException {
         }
     }
 
+    public synchronized void clean() {
+        StreamsException firstException = null;
+
+        try {
+            cleanRemovedTasks(0, false);
+        } catch (IOException e) {
+            // this is already logged within cleanRemovedTasks
+            firstException = new StreamsException(e);
+        }
+        try {
```

Comment:
Should we still continue deleting the global state directory if the local task directories cleanup failed? I'm thinking if the local stores deletion fails while global store succeeded, we will throw an exception, and user may retry in which case `globalStateDir()` will be called and trying to re-create the folder and delete it.

Reply:
We also "keep going" of cleaning a task directory fails (ie, don't stop on first error). Thus, it's a consistent pattern. We could of course change this via a "failFast" flag for `cleanRemovedTasks`. WDYT?

Reply:
The difference is that in `cleanRemovedTasks()` (originally we do not throw exception) we do not return a boolean or throwing exception indicating if it completely succeeds or not. So this call itself is effectively "best effort" but we do not comment it probably since it is an internal function. Now as we change it to throw exception if parameterized so I'm just wondering if we should still make this behavior.

Anyways, it is just a wild thought but not really a suggested comment.

Reply:
It's just about code sharing. If we call it internally, we can keep "best effort" approach. If users calls it, we can apply fail-fast IMHO. Will update the code.